File: frontend/src/api/screener.ts
import { apiClient } from "@/api/client";

export type Verdict = "BUY" | "HOLD" | "AVOID";
export type Confidence = "HIGH" | "MEDIUM" | "LOW";

export interface ScreenerScores {
  value: number | null;
  quality: number | null;
  growth: number | null;
  technical: number | null;
}

export type TechnicalVerdict = "STRONG_BUY" | "BUY" | "NEUTRAL" | "SELL" | "STRONG_SELL";

export interface ScreenerRating {
  symbol: string;
  name: string | null;
  sector: string | null;
  verdict: Verdict;
  confidence: Confidence;
  composite: number;
  scores: ScreenerScores;
  data_completeness: number;
  ltp: number | null;
  pct_from_52w_high: number | null;
  pct_from_52w_low: number | null;
  ta_verdict: TechnicalVerdict | null;
  ta_score: number | null;
  notes: string | null;
  as_of: string | null;
}

export interface TaGauge {
  label: string;
  buy: number;
  neutral: number;
  sell: number;
  verdict: TechnicalVerdict;
}

export interface TechnicalRatingRow extends ScreenerRating {
  ta_detail: {
    verdict: TechnicalVerdict;
    score: number;
    gauges: { moving_average: TaGauge; technical_indicators: TaGauge; overall: TaGauge };
    signals: Record<string, number>;
    readings: Record<string, number | null>;
  };
}

export interface TechnicalRatingsResponse {
  available: boolean;
  reason?: string;
  as_of?: string | null;
  summary: {
    strong_buy: number;
    buy: number;
    neutral: number;
    sell: number;
    strong_sell: number;
    total: number;
  };
  sectors?: string[];
  last_run: ScreenerRun | null;
  ratings: TechnicalRatingRow[];
}

export interface ScreenerRun {
  started_at: string | null;
  finished_at: string | null;
  trigger: string;
  universe_size: number;
  scored: number;
  buy: number;
  hold: number;
  avoid: number;
  error: string | null;
}

export interface ScreenerRatingsResponse {
  available: boolean;
  reason?: string;
  as_of?: string | null;
  summary: { buy: number; hold: number; avoid: number; total: number };
  sectors?: string[];
  last_run: ScreenerRun | null;
  ratings: ScreenerRating[];
}

export interface FactorRow {
  metric: string;
  raw: number | null;
  score: number | null;
}

export interface ScreenerRatingDetail extends ScreenerRating {
  available: boolean;
  reason?: string;
  metrics: Record<string, number | null>;
  factors: {
    peer_basis: "sector" | "universe";
    peer_count: number;
    weights: Record<string, number>;
    value: FactorRow[];
    quality: FactorRow[];
    growth: FactorRow[];
    technical: FactorRow[];
  };
}

export const DEEP_DIVE_SECTIONS = [
  ["business_model", "Business model"],
  ["quarterly_results", "Latest quarterly results"],
  ["balance_sheet", "Balance sheet health"],
  ["competitive_position", "Competitive position"],
  ["management_quality", "Management quality"],
  ["technical_setup", "Technical setup"],
  ["catalysts", "Key upcoming catalysts"],
  ["bull_case", "Bull case"],
  ["bear_case", "Bear case"],
  ["valuation", "Valuation"],
  ["verdict", "Final verdict"],
] as const;

export interface ScreenerDeepDive {
  available: boolean;
  symbol: string;
  reason?: string;
  hint?: string;
  raw?: string;
  verdict?: string | null;
  model?: string | null;
  sections?: Record<string, string>;
  facts?: Record<string, unknown>;
  generated_at?: string | null;
  stale?: boolean;
}

export const screenerApi = {
  ratings: (params: { verdict?: Verdict; sector?: string; sort?: string } = {}) =>
    apiClient
      .get<ScreenerRatingsResponse>("/screener/ratings", { params })
      .then((r) => r.data),
  rating: (symbol: string) =>
    apiClient.get<ScreenerRatingDetail>(`/screener/ratings/${encodeURIComponent(symbol)}`).then((r) => r.data),
  technicalRatings: (params: { verdict?: TechnicalVerdict; sector?: string; sort?: string } = {}) =>
    apiClient
      .get<TechnicalRatingsResponse>("/screener/technical-ratings", { params })
      .then((r) => r.data),
  deepDive: (symbol: string, refresh = false) =>
    apiClient
      .get<ScreenerDeepDive>(`/screener/ratings/${encodeURIComponent(symbol)}/deepdive`, {
        params: refresh ? { refresh: true } : {},
      })
      .then((r) => r.data),
};

File: backend/app/models/screener.py
"""Persistence for the Stock Screener.

Once a day :mod:`app.screener.engine` sweeps the liquid NSE universe,
pulls fundamentals (configured provider) + technicals (Kite daily
candles), cross-sectionally ranks every name and writes one
:class:`ScreenerRating` per stock. The rating is a transparent composite
of four 0-100 pillars (value / quality / growth / technical) plus the raw
inputs that fed them, bucketed BUY / HOLD / AVOID by score. A name whose
data is too thin is forced to HOLD and flagged low-confidence.

Nothing here is advice. It is a screen with its full working shown.
"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, Float, Index, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin


class ScreenerRating(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "screener_ratings"

    symbol: Mapped[str] = mapped_column(String(32), nullable=False, unique=True, index=True)
    name: Mapped[str | None] = mapped_column(String(200))
    sector: Mapped[str | None] = mapped_column(String(64), index=True)

    verdict: Mapped[str] = mapped_column(String(8), nullable=False, index=True)  # BUY | HOLD | AVOID
    confidence: Mapped[str] = mapped_column(String(8), nullable=False)  # HIGH | MEDIUM | LOW

    composite: Mapped[float] = mapped_column(Float, nullable=False, index=True)
    value_score: Mapped[float | None] = mapped_column(Float)
    quality_score: Mapped[float | None] = mapped_column(Float)
    growth_score: Mapped[float | None] = mapped_column(Float)
    technical_score: Mapped[float | None] = mapped_column(Float)
    data_completeness: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)

    ltp: Mapped[float | None] = mapped_column(Float)
    pct_from_52w_high: Mapped[float | None] = mapped_column(Float)
    pct_from_52w_low: Mapped[float | None] = mapped_column(Float)

    # TradingView-style technical rating from the same daily candles
    # (STRONG_BUY | BUY | NEUTRAL | SELL | STRONG_SELL); ta_score is the
    # net buy-minus-sell ratio in [-1, 1]; ta_detail holds the 3 gauges +
    # the 9 raw signals. Null when there were too few candles.
    ta_verdict: Mapped[str | None] = mapped_column(String(12), index=True)
    ta_score: Mapped[float | None] = mapped_column(Float)
    ta_detail: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)

    # every raw metric + per-pillar sub-factor breakdown, as pulled
    metrics: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    factors: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    notes: Mapped[str | None] = mapped_column(Text)

    as_of: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    __table_args__ = (Index("ix_screener_verdict_score", "verdict", "composite"),)


class ScreenerDeepDive(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    """A generated, on-demand narrative deep-dive for one stock. The
    sections are written by the configured assistant LLM strictly from the
    numbers the screener already pulled — anything unsupported is left as
    "Not enough data." Cached per symbol; regenerated on request or when
    stale."""

    __tablename__ = "screener_deep_dives"

    symbol: Mapped[str] = mapped_column(String(32), nullable=False, unique=True, index=True)
    verdict: Mapped[str | None] = mapped_column(String(8))
    model: Mapped[str | None] = mapped_column(String(120))
    sections: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    facts: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    error: Mapped[str | None] = mapped_column(Text)
    generated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)


class ScreenerRun(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "screener_runs"

    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    trigger: Mapped[str] = mapped_column(String(16), nullable=False, default="schedule")
    universe_size: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    scored: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    buy: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    hold: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    avoid: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    error: Mapped[str | None] = mapped_column(Text)
